### PR TITLE
[Feat] Optimize Eagle3

### PR DESCRIPTION
**Pull request by lingebeng**
## Summary

This PR optimizes EAGLE3 speculative decoding by fusing the recurrent multi-step draft generation and target verification into a single JIT, and enables the overlap scheduler for EAGLE3 (previously only NEXTN and DFLASH were supported).

### What changed

- **Fused recurrent EAGLE3 draft**: `_build_eagle3_recurrent_draft_extend` runs the draft-extend (step 0) plus all recurrent decode steps in one JIT on device, eliminating per-step host round-trips.
- **Fused EAGLE3 verify**: the linear-chain verify path now maps the draft-vocabulary chain to target tokens (`hot_token_ids`) inside the fused verify JIT, and rebuilds dynamic FA attention metadata on device from a page-only base metadata upload, cutting H2D transfers.
- **EAGLE3 overlap scheduling**: the scheduler now supports EAGLE3 (topk=1, `num_draft_tokens == num_steps + 1`, FA backend) on the spec-decode overlap path. New prefills joining an active decode batch are parked for one round so their width-1 bootstrap state can complete its first decode and publish req-indexed relay state before merging.
- **EAGLE state refactor**: `EagleDraftInput` / `EagleVerifyInput` moved to a new `eagle_info.py`; removed the `EagleVerifyOutput` dataclass, the `pending_draft_extend_result` relay mechanism, and unused pytree fields (Draft children 13→10, Verify 9→6). `eagle_util.py` keeps only tree construction and shared KV-mapping helpers. `merge_batch` now handles mixing full recurrent chains with width-1 prefill seeds by downgrading to the common seed.
- **Perf micro-optimizations**: no-overlap decode copies only the host arrays actually consumed, reuses the `cache_loc` buffer, caches verify placeholders and greedy sampling constants, and `gather_spec_relay_buffers` restores flat data sharding.
- **Docs**: updated `docs/architecture/09-speculative-decoding.md`; added ~440 lines of unit tests covering the new fused paths, sharding, metadata builders, and relay gather.

### Notes

- The draft model now keeps its own head (embedding sharing only) instead of replacing it with a hot-token-remapped copy of the target head; GSM8K accuracy is unchanged (0.9242 vs 0.9249, within noise).
- `SGL_JAX_DISABLE_FUSED_EAGLE3_RECURRENT_DRAFT=1` is provided as an escape hatch to fall back to the legacy draft loop.

## Profile
- qwen3-8B-eagle3
- tpuv6e-4
- batch_size = 16
- input/output = 1024/4096
### Before

image

### After

image

### Accuracy Benchmark

> evalscope v1.9.0

Mode | Dataset | Accuracy | num
-- | -- | -- | --
No Eagle3 | GSM8K | 0.9249 | 1220/1319
Eagle3 | GSM8K | 0.9242 | 1219/1319

```bash
evalscope eval \
    --model /models/Qwen/Qwen3-8B \
    --eval-type openai_api \
    --api-url http://127.0.0.1:30000/v1/chat/completions \
    --datasets gsm8k \
    --eval-batch-size 64 \
    --generation-config '{"max_tokens":4096,"temperature":0.0}' \
    --work-dir /tmp/evalscope_results
```
### Speed Benchmark

C | Eagle3 Overlap | Without Eagle3 Overlap | Speedup | Eagle3 No-overlap | Without Eagle3 No-overlap | Speedup
-- | -- | -- | -- | -- | -- | --
1 | 666.9 | 178.1 | 3.74× | 444.9 | 137.3 | 3.24×
2 | 1,365.1 | 344.3 | 3.96× | 506.8 | 271.8 | 1.86×
4 | 1,864.8 | 663.3 | 2.81× | 623.0 | 522.4 | 1.19×
8 | 4,235.5 | 1,294.6 | 3.27× | 1,596.7 | 962.6 | 1.66×
16 | 5,360.8 | 2,481.7 | 2.16× | 2,551.7 | 1,659.5 | 1.54×
32 | 6,665.0 | 4,091.5 | 1.63× | 3,808.5 | 2,726.6 | 1.40×
64 | 7,434.4 | 5,031.4 | 1.48× | 4,483.8 | 3,883.9 | 1.15×
128 | 7,869.0 | 5,704.8 | 1.38× | 5,477.6 | 4,008.8 | 1.37×

image

#### Hardware

- **TPU:** v6e, 4 chips (`v6e-4` slice)
- **Topology:** `2x2`
- **JAX devices:** 4
- **Parallelism:** `--tp-size 4 --dp-size 1` (TP=4, DP=1)

#### Software Versions

- **sglang-jax:** `0.0.0.dev1+gcbe262e59`
- **Commit:** `21291484e` (PR #1534 head + 2 sharding fixes)
- **JAX:** `0.10.2`
- **jaxlib:** `0.10.2`
- **libtpu:** `0.0.42.1`
- **Python:** `3.12.12`

#### Models

- **Target model:** `/models/Qwen3-8B`
  - dtype: `bfloat16`
  - Storage: GCS mount
- **Draft model:** `thoughtworks/Qwen3-8B-Eagle3`

#### Dataset (`bench_serving`)

- **Dataset:** `random`
- **Input length:** 1024 tokens
- **Output length:** 4096 tokens
- **Range ratio:** `1.0`
- **Num prompts:** `3 × C`
  - e.g. `C=128` → 384 prompts
- **Concurrency:** `1 / 2 / 4 / 8 / 16 / 32 / 64 / 128`
- **Request rate:** `inf`
- **Radix cache:** disabled
  - Cache hit rate = 0
  - Prefill is not cache-served

#### Server Launch Commands

Four configurations are evaluated:

1. Eagle3 + overlap
2. Eagle3 + no-overlap
3. Without Eagle3 + overlap
4. Without Eagle3 + no-overlap

The no-overlap variants add:

```bash
--disable-overlap-schedule
```

#### Eagle3 (Overlap)

```bash
./.venv/bin/python -m sgl_jax.launch_server \
  --model-path /models/Qwen3-8B \
  --speculative-algorithm EAGLE3 \
  --speculative-draft-model-path thoughtworks/Qwen3-8B-Eagle3 \
  --speculative-num-steps 3 \
  --speculative-num-draft-tokens 4 \
  --speculative-eagle-topk 1 \
  --tp-size 4 \
  --dp-size 1 \
  --dtype bfloat16 \
  --attention-backend fa \
  --mem-fraction-static 0.65 \
  --page-size 64 \
  --chunked-prefill-size 2048 \
  --max-running-requests 128 \
  --trust-remote-code \
  --disable-radix-cache \
  --grammar-backend none \
  --host 0.0.0.0 \
  --port 30000
```

#### Without Eagle3 (Overlap)

```bash
./.venv/bin/python -m sgl_jax.launch_server \
  --model-path /models/Qwen3-8B \
  --tp-size 4 \
  --dp-size 1 \
  --dtype bfloat16 \
  --attention-backend fa \
  --mem-fraction-static 0.65 \
  --page-size 64 \
  --chunked-prefill-size 2048 \
  --max-running-requests 128 \
  --trust-remote-code \
  --disable-radix-cache \
  --grammar-backend none \
  --host 0.0.0.0 \
  --port 30000
```
#### Benchmark Command

For each concurrency `C`:

```bash
./.venv/bin/python -m sgl_jax.bench_serving \
  --backend sgl-jax \
  --base-url http://127.0.0.1:30000 \
  --model /models/Qwen3-8B \
  --tokenizer /models/Qwen3-8B \
  --dataset-name random \
  --random-input-len 1024 \
  --random-output-len 4096 \
  --random-range-ratio 1.0 \
  --num-prompts $((C*3)) \
  --max-concurrency "$C" \
  --disable-tqdm \
  --output-file "$OUT/bs_$C/result.jsonl"
```

### CODE REVIEW

**Review by cjx0709 — APPROVED** [review]
LGTM
